File: src/js/4_BehavioralPatterns/7_Observer/property_dependencies.js
class Event
{
  constructor()
  {
    this.handlers = new Map();
    this.count = 0;
  }

  subscribe(handler)
  {
    this.handlers.set(++this.count, handler);
    return this.count;
  }

  unsubscribe(idx)
  {
    this.handlers.delete(idx);
  }

  // 1) who fired the event?
  // 2) additional data (event args)
  fire(sender, args)
  {
    this.handlers.forEach(
      (v, k) => v(sender, args)
    );
  }
}

class Person
{
  constructor(age)
  {
    this._age = age;
    this.property_changed = new Event();
  }

  get age() {
    return _age;
  }

  set age(value)
  {
    if (!value || this._age === value)
      return; // nothing to do here

    let oldCanVote = this.canVote;

    this._age = value;
    this.property_changed.fire(
      this,
      new PropertyChangedArgs('age', value)
    );

    if (oldCanVote !== this.canVote)
    {
      this.property_changed.fire(
        this,
        new PropertyChangedArgs('canVote', this.canVote)
      );
    }
  }

  get canVote()
  {
    return this._age >= 16;
  }
}

class PropertyChangedArgs
{
  constructor(name, newValue)
  {
    this.name = name;
    this.newValue = newValue;
  }
}

class VotingChecker
{
  constructor(person) {
    this.person = person;
    this.person.property_changed.subscribe(
      this.voting_changed.bind(this));
  }

  voting_changed(sender, args)
  {
    if (sender === this.person && args.name === 'canVote')
    {
      console.log('Voting status changed to ' + args.newValue);
    }
  }
}

let person = new Person('John');
let checker = new VotingChecker(person);
for (let i = 10; i < 20; ++i)
{
  console.log(`Changing age to ${i}`);
  person.age = i;
}

File: src/go/4_BehavioralPatterns/7_Observer/propertydependencies.go
package observer

import (
  "container/list"
  "fmt"
)

type Observable struct {
  subs *list.List
}

func (o *Observable) Subscribe(x Observer) {
  o.subs.PushBack(x)
}

func (o *Observable) Unsubscribe(x Observer) {
  for z := o.subs.Front(); z != nil; z = z.Next() {
    if z.Value.(Observer) == x {
      o.subs.Remove(z)
    }
  }
}

func (o *Observable) Fire(data interface{}) {
  for z := o.subs.Front(); z != nil; z = z.Next() {
    z.Value.(Observer).Notify(data)
  }
}

type Observer interface {
  Notify(data interface{})
}

type Person struct {
  Observable
  age int
}

func NewPerson(age int) *Person {
  return &Person{Observable{new(list.List)}, age}
}

type PropertyChanged struct {
  Name string
  Value interface{}
}

func (p *Person) Age() int { return p.age }
func (p *Person) SetAge(age int) {
  if age == p.age { return } // no change

  oldCanVote := p.CanVote()

  p.age = age
  p.Fire(PropertyChanged{"Age", p.age})

  if oldCanVote != p.CanVote() {
    p.Fire(PropertyChanged{"CanVote", p.CanVote()})
  }
}

func (p *Person) CanVote() bool {
  return p.age >= 18
}

type ElectrocalRoll struct {
}

func (e *ElectrocalRoll) Notify(data interface{}) {
  if pc, ok := data.(PropertyChanged); ok {
    if pc.Name == "CanVote" && pc.Value.(bool) {
      fmt.Println("Congratulations, you can vote!")
    }
  }
}

func main() {
  p := NewPerson(0)
  er := &ElectrocalRoll{}
  p.Subscribe(er)

  for i := 10; i < 20; i++ {
    fmt.Println("Setting age to", i)
    p.SetAge(i)
  }
}